### ISSUE TITLE
browser-esbuild messes up resolution of scss imports, adding tildes

### ISSUE THREAD

**Issue by LorinRenodeyn** [angular/angular-cli#27272]
### Command

serve

### Is this a regression?

- [X] Yes, this behavior used to work in the previous version

### The previous version in which this bug was not present was

17.1.x

### Description

When switching from `@angular-devkit/build-angular:browser` to `@angular-devkit/build-angular:browser-esbuild`, the ng-serve command throws errors on the scss imports. However, the errors logged to the terminal don't correspond to the code of the file the logs refer to. It appears that something is going wrong with the import resolution of the scss files.

### Minimal Reproduction

Command run: `yarn ng serve` (with angular.json configured to `browser-esbuild`) or `yarn ng serve --force-esbuild`

This is in an angular workspace with an application (named 'application') and  2 library packages (/packages/components and /packages/icons). The console errors refer to the scss base file in /application/src/styles.scss, but the logs don't correspond to the file contents. The imported packages are internal npm packages used by the client and names have been redacted.

styles.scss contents:
```SCSS
@import '@component-lib/styles/reset';
@import '@component-lib/styles/define-colors';

@import '@component-lib/styles/colors';
@import '@component-lib/styles/typography';

@import '@component-lib/styles/fonts/regular';

body {
  @include style-lib-font-normal;

  background-color: $style-lib-color-base-default;
  overscroll-behavior: none;
}

```

Running the exact same codebase with the `@angular-devkit/build-angular:browser` config works fine and without errors, both in terminal and browser console.The tildes reported by the error logs are _not_ present in the codebase.

This is a new project with only the boilerplate setup provided by the client.

[edit, adding config info from angular.json]

```JSON
"application": {
[...]
    "build": {
    "builder": "@angular-devkit/build-angular:browser",
    "options": {
      "assets": ["application/src/favicon.ico", "application/src/assets"],
      "buildOptimizer": false,
      "extractLicenses": false,
      "index": "application/src/index.html",
      "inlineStyleLanguage": "scss",
      "main": "application/src/main.ts",
      "namedChunks": true,
      "optimization": false,
      "outputPath": "dist/application",
      "polyfills": "application/src/polyfills.ts",
      "scripts": [],
      "sourceMap": true,
      "styles": ["application/src/styles.scss"],
      "tsConfig": "application/tsconfig.app.json",
      "vendorChunk": true
    },
    "configurations": {
      "de": {
        "baseHref": "/",
        "localize": ["de"]
      },
      "e2e": {
        "fileReplacements": [
          {
            "replace": "application/src/environments/environment.ts",
            "with": "application/src/environments/environment.e2e.ts"
          }
        ]
      },
      "en": {
        "baseHref": "/",
        "localize": ["en"]
      },
      "fr": {
        "baseHref": "/",
        "localize": ["fr"]
      },
      "nl": {
        "baseHref": "/",
        "localize": ["nl"]
      },
      "no-development": {
        "tsConfig": "application/tsconfig.app.no-dev.json"
      },
    }
  },
[...]
  "serve": {
          "builder": "@angular-devkit/build-angular:dev-server",
          "options": {
            "buildTarget": "application:build"
          },
          "configurations": {
            "de": {
              "buildTarget": "application:build:de"
            },
            "e2e": {
              "buildTarget": "application:build:e2e"
            },
            "e2e-no-development": {
              "buildTarget": "application:build:e2e,no-development"
            },
            "en": {
              "buildTarget": "application:build:en"
            },
            "fr": {
              "buildTarget": "application:build:fr"
            },
            "nl": {
              "buildTarget": "application:build:nl"
            },
            "no-development": {
              "buildTarget": "application:build:no-development"
            },
            "production": {
              "buildTarget": "application:build:production"
            }
          }
        },
```

### Exception or Error

```text
✘ [ERROR] Could not resolve "~normalize.css/normalize.css"

    application/src/styles.scss:1:8:
      1 │ @import '~normalize.css/normalize.css';
        ╵         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  You can mark the path "~normalize.css/normalize.css" as external to exclude it from the bundle, which will remove this error and leave the unresolved path in the bundle.


✘ [ERROR] Could not resolve "~@component-lib/styles/fonts/museo/museosans-300-webfont.woff2" [plugin angular-css-resource]

    application/src/styles.scss:14:11:
      14 │   src: url("~@component-lib/styles/fonts/museo/museosans-300-webfont.woff2"...
         ╵            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  You can remove the tilde and use a relative path to reference it, which should remove this error.
  Preprocessor stylesheets may not show the exact file location of the error.


✘ [ERROR] Could not resolve "~@component-lib/styles/fonts/museo/museosans-300-webfont.woff" [plugin angular-css-resource]

    application/src/styles.scss:14:89:
      14 │ ...f2"), url("~@component-lib/styles/fonts/museo/museosans-300-webfont.woff...
         ╵              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  You can remove the tilde and use a relative path to reference it, which should remove this error.
  Preprocessor stylesheets may not show the exact file location of the error.


✘ [ERROR] Could not resolve "~@component-lib/styles/fonts/museo/museosans-500-webfont.woff2" [plugin angular-css-resource]

    application/src/styles.scss:26:11:
      26 │   src: url("~@component-lib/styles/fonts/museo/museosans-500-webfont.woff2"...
         ╵            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  You can remove the tilde and use a relative path to reference it, which should remove this error.
  Preprocessor stylesheets may not show the exact file location of the error.


✘ [ERROR] Could not resolve "~@component-lib/styles/fonts/museo/museosans-500-webfont.woff" [plugin angular-css-resource]

    application/src/styles.scss:26:89:
      26 │ ...f2"), url("~@component-lib/styles/fonts/museo/museosans-500-webfont.woff...
         ╵              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  You can remove the tilde and use a relative path to reference it, which should remove this error.
  Preprocessor stylesheets may not show the exact file location of the error.


✘ [ERROR] Could not resolve "~@component-lib/styles/fonts/museo/museosans-700-webfont.woff2" [plugin angular-css-resource]

    application/src/styles.scss:38:11:
      38 │   src: url("~@component-lib/styles/fonts/museo/museosans-700-webfont.woff2"...
         ╵            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  You can remove the tilde and use a relative path to reference it, which should remove this error.
  Preprocessor stylesheets may not show the exact file location of the error.


✘ [ERROR] Could not resolve "~@component-lib/styles/fonts/museo/museosans-700-webfont.woff" [plugin angular-css-resource]

    application/src/styles.scss:38:89:
      38 │ ...f2"), url("~@component-lib/styles/fonts/museo/museosans-700-webfont.woff...
         ╵              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  You can remove the tilde and use a relative path to reference it, which should remove this error.
  Preprocessor stylesheets may not show the exact file location of the error.


✘ [ERROR] Could not resolve "~@component-lib/styles/fonts/museo/roboto-mono-xxx-light-webfont.woff2" [plugin angular-css-resource]

    application/src/styles.scss:44:11:
      44 │   src: url("~@component-lib/styles/fonts/museo/roboto-mono-xxx-light-webfon...
         ╵            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  You can remove the tilde and use a relative path to reference it, which should remove this error.
  Preprocessor stylesheets may not show the exact file location of the error.


✘ [ERROR] Could not resolve "~@component-lib/styles/fonts/museo/roboto-mono-xxx-light-webfont.woff" [plugin angular-css-resource]

    application/src/styles.scss:44:97:
      44 │ ...f2"), url("~@component-lib/styles/fonts/museo/roboto-mono-xxx-light-webf...
         ╵              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  You can remove the tilde and use a relative path to reference it, which should remove this error.
  Preprocessor stylesheets may not show the exact file location of the error.


✘ [ERROR] Could not resolve "~@component-lib/styles/fonts/museo/roboto-mono-xxx-regular-webfont.woff2" [plugin angular-css-resource]

    application/src/styles.scss:50:11:
      50 │   src: url("~@component-lib/styles/fonts/museo/roboto-mono-xxx-regular-webf...
         ╵            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  You can remove the tilde and use a relative path to reference it, which should remove this error.
  Preprocessor stylesheets may not show the exact file location of the error.


✘ [ERROR] Could not resolve "~@component-lib/styles/fonts/museo/roboto-mono-xxx-regular-webfont.woff" [plugin angular-css-resource]

    application/src/styles.scss:50:99:
      50 │ ...f2"), url("~@component-lib/styles/fonts/museo/roboto-mono-xxx-regular-we...
         ╵              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  You can remove the tilde and use a relative path to reference it, which should remove this error.
  Preprocessor stylesheets may not show the exact file location of the error.
```

### Your Environment

```text
_                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 17.2.3
Node: 18.14.0
Package Manager: yarn 3.6.0
OS: darwin arm64

Angular: 17.2.4
... common, compiler, compiler-cli, core, localize

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1702.3
@angular-devkit/build-angular   17.2.3
@angular-devkit/core            17.2.3
@angular-devkit/schematics      17.2.3
@angular/cdk                    17.2.2
@angular/cli                    17.2.3
@schematics/angular             17.2.3
rxjs                            7.8.1
typescript                      5.2.2
zone.js                         0.14.4
```

**Comment by Alijavedofficial** [comment]
If you're using a custom build configuration or a specific plugin (like angular-css-resource), ensure that it's configured to correctly resolve paths with the tilde symbol. You might need to adjust the configuration or use a different method to import your styles and fonts.

**Comment by LorinRenodeyn** [comment]
For clarification (I'll update the original report above too), the tilde symbols are _not_ present in the codebase.
I will add more detailed config info to the report as well.

tl;dr without any changes to the codebase:
`ng serve` => works fine
`ng server --force-esbuild` => errors

**Comment by Alijavedofficial** [comment]
Ensure that your angular.json file is correctly configured for the serve command with forceEsbuild: true.

**Comment by LorinRenodeyn** [comment]
According to the [angular docs](https://angular.io/guide/esbuild#using-the-browser-esbuild-builder), no changes beyond the builder are needed for esbuild to work:

> Changing the builder field is the only change you will need to make.
